File: synora-web/src/lib/security-mode.ts
import { isRecord } from "./normalize";

export type SecurityMode = "home" | "night" | "away" | "high_security";
export type ExpectedOccupancy = "unknown" | "occupied" | "empty";

export type SecurityModeState = {
  mode: SecurityMode;
  armed: boolean;
  expected_occupancy: ExpectedOccupancy;
  set_by: string;
  reason: string;
  since: string | null;
  expires_at: string | null;
  source: string;
};

export function normalizeSecurityMode(value: unknown): SecurityModeState {
  const source = isRecord(value) ? value : {};
  const rawMode = typeof source.mode === "string" ? source.mode.trim().toLowerCase() : "home";
  const mode: SecurityMode = rawMode === "night" || rawMode === "away" || rawMode === "high_security" ? rawMode : "home";
  const occupancy: ExpectedOccupancy = mode === "night"
    ? "occupied"
    : mode === "away" || mode === "high_security"
      ? "empty"
      : source.expected_occupancy === "occupied" || source.expected_occupancy === "empty" ? source.expected_occupancy : "unknown";
  return {
    mode,
    armed: mode !== "home" && source.armed !== false,
    expected_occupancy: occupancy,
    set_by: typeof source.set_by === "string" ? source.set_by : "system",
    reason: typeof source.reason === "string" ? source.reason : "",
    since: typeof source.since === "string" ? source.since : null,
    expires_at: typeof source.expires_at === "string" ? source.expires_at : null,
    source: typeof source.source === "string" ? source.source : "system",
  };
}

export function securityModeLabel(mode: SecurityMode): string {
  return { home: "Maison", night: "Nuit", away: "Absent", high_security: "Sécurité élevée" }[mode];
}

File: pkg/contract/security_mode.go
package contract

import "time"

type SecurityMode string

const (
	SecurityModeHome         SecurityMode = "home"
	SecurityModeNight        SecurityMode = "night"
	SecurityModeAway         SecurityMode = "away"
	SecurityModeHighSecurity SecurityMode = "high_security"
)

type ExpectedOccupancy string

const (
	ExpectedOccupancyUnknown  ExpectedOccupancy = "unknown"
	ExpectedOccupancyOccupied ExpectedOccupancy = "occupied"
	ExpectedOccupancyEmpty    ExpectedOccupancy = "empty"
)

type SecurityModeState struct {
	Mode              SecurityMode      `json:"mode"`
	Armed             bool              `json:"armed"`
	ExpectedOccupancy ExpectedOccupancy `json:"expected_occupancy"`
	SetBy             string            `json:"set_by"`
	Reason            string            `json:"reason"`
	Since             time.Time         `json:"since"`
	ExpiresAt         *time.Time        `json:"expires_at"`
	Source            string            `json:"source"`
}

type SecurityModeRequest struct {
	Mode            SecurityMode `json:"mode"`
	Reason          string       `json:"reason,omitempty"`
	DurationSeconds int          `json:"duration_seconds,omitempty"`
	SetBy           string       `json:"set_by,omitempty"`
	Source          string       `json:"source,omitempty"`
}

type SecurityArmRequest = SecurityModeRequest

type SecurityDisarmRequest struct {
	Reason string `json:"reason,omitempty"`
	SetBy  string `json:"set_by,omitempty"`
	Source string `json:"source,omitempty"`
}

func DefaultSecurityModeState(now time.Time) SecurityModeState {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	return SecurityModeState{Mode: SecurityModeHome, ExpectedOccupancy: ExpectedOccupancyUnknown, SetBy: "system", Reason: "default", Since: now.UTC(), Source: "system"}
}

func NormalizeSecurityModeState(value SecurityModeState, now time.Time) SecurityModeState {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	if value.Mode == "" {
		value.Mode = SecurityModeHome
	}
	switch value.Mode {
	case SecurityModeHome:
		value.Armed = false
		if value.ExpectedOccupancy == "" {
			value.ExpectedOccupancy = ExpectedOccupancyUnknown
		}
	case SecurityModeNight:
		value.Armed = true
		value.ExpectedOccupancy = ExpectedOccupancyOccupied
	case SecurityModeAway, SecurityModeHighSecurity:
		value.Armed = true
		value.ExpectedOccupancy = ExpectedOccupancyEmpty
	default:
		value.Mode = SecurityModeHome
		value.Armed = false
		value.ExpectedOccupancy = ExpectedOccupancyUnknown
	}
	if value.SetBy == "" {
		value.SetBy = "system"
	}
	if value.Source == "" {
		value.Source = "system"
	}
	if value.Since.IsZero() {
		value.Since = now.UTC()
	}
	return value
}
